### ISSUE TITLE
Append, Prepend, Replace - dont seem to work

### ISSUE THREAD

**Issue by DannyBen** [piotrmurach/tty-file#3]
Hi,

Trying this lovely gem on Ubuntu 16, with this script:

```ruby
# test.rb
require 'tty-file'

# replace me

tty = TTY::File
tty.inject_into_file 'test.rb', "# injected\n\n", after: "TTY::File\n"
tty.replace_in_file 'test.rb', /replace me/, "replaced"
tty.prepend_to_file 'test.rb', "# prepended"
tty.append_to_file 'test.rb', "# appended"
```

Only the inject works, the rest wont.

Here is a screencast:
[![asciicast](https://asciinema.org/a/Mdj27wKvV2QgfibuLJIHygQf1.png)](https://asciinema.org/a/Mdj27wKvV2QgfibuLJIHygQf1)

Not sure what I am doing wrong.

Ruby version: ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux]

**Comment by piotrmurach** [comment]
Hey,

Sorry this issue has escaped my attention. What you're trying to do is a bit unusual in a sense that I haven't aimed these operations at modifying the file they are defined in, it's quite meta!

Given `another.rb` file:
```ruby
# replace me
TTY::File
```
and then given your file targeting `another.rb` 

```ruby
# test.rb
require 'tty-file'

# replace me

tty = TTY::File
tty.inject_into_file 'another.rb', "# injected\n\n", after: "TTY::File\n"
tty.replace_in_file 'another.rb', /replace me/, "replaced"
tty.prepend_to_file 'another.rb', "# prepended"
tty.append_to_file 'another.rb', "# appended"
```
The output is as expected:
```ruby
#prepended#replaced
TTY::File
#injected

#appended
```

**Comment by DannyBen** [comment]
Huh. Never thought this would be the cause...

**Comment by piotrmurach** [comment]
Ahh ok... the reason why it doesn't work is that there is a check to see if the content that you wish to replace/append/prepend already exists. The `# prepanded`, `# appended`, and `replaced` are already present in the source file, hence nothing is done. Therefore, using the `:force` option makes this work:

```ruby
# test.rb
require 'tty-file'

# replace me

tty = TTY::File
tty.inject_into_file 'test.rb', "# injected\n\n", after: "TTY::File\n"
tty.replace_in_file 'test.rb', /replace me/, "replaced", force: true
tty.prepend_to_file 'test.rb', "# prepended", force: true
tty.append_to_file 'test.rb', "# appended", force: true
```

Having said that, I'm thinking about removing this check, since appending, prepending is an operation that should run ad infinitum. I will only raise error if you wish to replace content with regex that doesn't match anything in file. However, you will always be able to force it. Does that sound reasonable to you?

**Comment by piotrmurach** [comment]
It's all in master if you have time to give it a test.

**Comment by DannyBen** [comment]
Hmm. Just did, and I dont see any change in behavior.

Using tty-file 0.6.0 (was 0.5.0) from https://github.com/piotrmurach/tty-file (a
t master@f4f14a5)

It still does not "update itself" (but of course this is not important), and the `#prepend_to_file` and `#append_to_file` still avoid updating if they find a similar phrase in the target file.

**Comment by piotrmurach** [comment]
The `append_to_file` and `prepend_to_file` rely on `inject_into_file` for their implementation. The `inject_into_file` relies on `replace_in_file`. Therefore, the change in `replace_in_file` means that all the methods, provided the file contains the matching content will make a change. So you can append, prepend, inject ad infinitum.

Previously, you couldn't do it due to check being done if content is already present. I came to conclusion that the 'intelligence' of preventing file modification was too much and may lead to buggy code working correctly. At the end of the day, if you wish to continue appending same value then you should be able to. 

Does that sound ok with you or do you believe that injecting/appending/prepending should only work when the content is not present at the given location in source file?

**Comment by DannyBen** [comment]
That is an interesting dilemma, and my thoughts, in no particular order are there:

1. Initially, I did expect all the methods to be "dumb" and just do as they are told, when it did not happen, I thought it is a bug
2. When I learned that there is some intelligence here, I thought "well, this is useful...".
3. If this intelligence stays, I believe it should be precise. Meaning, if I try to append the line `hello world`, then it will be appended unless the line matches `/^hello world$/`
4. If both the intelligent and "dumb" method stay, one mode should be the default, and the other should happen with a flag.
5. If the "dumb" way stays the default, then perhaps a parameter such as `once: true` should allow the user to move to the intelligent way.
6. Everything I am saying is just since you already implemented both ways, which perhaps would be a shame to lose, since (maybe) some people would want one, while others will prefer the other.

In case my vote counts for anything, I would say this:

1. The default should be "dumb" - append, replace, prepend without checking content
2. If easy to implement - add a parameter `once: true` which will check for exact content before processing.
3. If not easy (== potential for bugs) to implement exact filter, still go with `once: true`, and make it work as the old intelligence worked.
4. If items 2 and 3 above are not sitting well with your vision, then perhaps indeed you are right, and the forced method should be the only one to prevail.

Makes sense?

**Comment by piotrmurach** [comment]
Thank you for your thoughtful feedback! Just got reminded of this issue as I need to publish `tty-file` update. 

What I think is the best course of action is to:
1. By default use the 'intelligent' behaviour and prevent overriding content if it already exists. This may prevent a class of bugs and provide better developer experience.
2. Use `:force` flag to allow overriding file changes as it is consistent with other methods and also means that user intends to replace content regardless. I prefer this to `:once` option which is inconsistent with other api calls and implies numerical idea to replacements.
3. I would agree with you that `append` and `prepend` should only check for respective file ending and beginning.

**Comment by piotrmurach** [comment]
I've decided to be more restrictive with detecting the presence of the replacement string. It only considers complete line matches which will prevent matching random content inside a file. Please [see commit](https://github.com/piotrmurach/tty-file/commit/ab46379adc17c3e22fd7f00bdb44d6b8b3d05a9e) for details.

> 1. Initially, I did expect all the methods to be "dumb" and just do as they are told, when it did not happen, I thought it is a bug

The current behaviour means that your `test.rb` example will work as expected.

> 2. When I learned that there is some intelligence here, I thought "well, this is useful...".
> 3. If this intelligence stays, I believe it should be precise. Meaning, if I try to append the line hello world, then it will be appended unless the line matches /^hello world$/

There is a 'middleground' which means there is a bit of 'intelligence' that, as you have suggested, only cosiders a full line for matching. The `:force` flag is used to keep on appending content.

**Comment by piotrmurach** [comment]
Released `v0.6.0` with the patch.
